NKL/Trace/Python.lean

```diff
@@ -0,0 +1,234 @@
+
+partial def stmt : Stmt -> Tracer Unit
+  | .stmtPos s' p => withPos p (stmt' s')
+
+partial def stmt' : Stmt' -> Tracer Unit
+  | .expr (.exprPos (.const _) _) => return ()
+  | .expr e => do
+      match <- term e with
+      | .expr e _ => add_stmt (.expr e)
+      | _ => return ()  -- effects are done, can be removed from KLR
+  | .assert e => do
+      let t <- term e
+      if (<- t.isFalse) then throw "assertion failed"
+  | .assign xs e => assign xs e
+  | .augAssign x op e => do
+      stmt' (.assign [x] (.exprPos (.binOp op x e) (<- getPos)))
+  | .annAssign _ _ .none => return ()
+  | .annAssign x _ (.some e) => stmt' (.assign [x] e)
+  | _s => throw "not yet implemented" --s!"unimp {repr s}"
+
+-- Bind positional and keyword arguments to a Python function.
+-- Note: default arguments should be evaluated in the global environment,
+-- however we know that each source function begins with an empty local
+-- environment, so it is OK to evaluate the default arguments in the
+-- functions initial environment.
+
+partial def bind_args (f : Fun)
+                      (args : List Term)
+                      (kwargs : List (String × Term))
+                      : Tracer (List (String × Term)) := do
+  if f.args.vararg != none || f.args.kwarg != none then
+    throw "var args not supported"
+  if args.length < f.args.posonlyargs.length then
+    throw "not enough arguments"
+  let dflts := f.args.all_defaults
+  let names := f.args.names
+  if args.length + kwargs.length > names.length then
+    throw "too many arguments supplied (varargs not supported)"
+  let argmap <- f.args.names.enum.mapM fun (i,x) => do
+    if h:args.length > i then
+      return (x, args.get (Fin.mk i h))
+    else if let some v := kwargs.lookup x then
+      return (x, v)
+    else if let some e := dflts.lookup x then
+      return (x, <- term' e)
+    else
+      throw s!"argument {x} not supplied"
+  return argmap
+
+-- For a function call, first evaluate the argument in the current environment.
+-- Then enter a new environment and evaluate the function statements.
+partial def function_call (f : Fun)
+                          (args : List Term)
+                          (kwargs : List (String × Term))
+                          : Tracer Unit := do
+  let args <- bind_args f args kwargs
+  let args <- args.mapM fun (x,e) => return (x, e)
+  withSrc f.source $ enterFun $ do
+    args.forM fun (x,e) => do extend x.toName e
+    f.body.forM stmt
+
+end
+
+-- Evaluate each global in the current environment, skipping any globals that
+-- are already defined. Note, we may have globals or functions with dummy
+-- implementations, e.g.
+--   def add(x,y): pass
+-- If we have an internal definition, we will use this over anything found
+-- during parsing.
+
+private def globals (k : Kernel) : Tracer Unit := do
+  let s <- get
+  for (n, f) in k.funcs do
+    let n := n.toName
+    if not (s.env.contains n) then
+      extend_global n (.source f)
```

Comment:
I can't find this definition. This updates the TraceM inner monad? Maybe I'll take a break and catch up when you merge some of the earlier PRs.